### DISCUSSION TITLE
Should we provide an "escape hatch" from WebTargetHelper?

### DISCUSSION THREAD

**Discussion by sleberknight** [kiwiproject/kiwi#599] [Ideas]
I just came across some code in our of our services which is using the JAX/Jakarta-RS `WebTarget`, and which could have used kiwi's `WebTargetHelper`. Basically, the code has a method that accepts a `WebTarget` and a `Map<String, Object>` of query parameters. It adds the given query parameters and returns a new `WebTarget` using code similar to what we've already written and put in `WebTargetHelper`.

So, in theory this code could easily have used `WebTarget#queryParamsFromMap` _except_ that the method in the service returns a `WebTarget` while `queryParamsFromMap` returns a `WebTargetHelper` for method chaining purposes. 

The problem is that there is no easy "escape hatch" from `WebTargetHelper` to simply return the current `WebTarget`. Since this method described above is intended to simply add query parameters to _any_ `WebTarget` and return the new `WebTarget` with those parameters added, and then let the caller do whatever it likes, this won't work.

So the question is whether we should provide an "escape hatch" to permit this use case. Example code:

```java
// This is what the utility method looked like:
static WebTarget targetWithQueryParams(WebTarget webTarget, Map<String, Object> parameters) {
   // ...bunch of code to create a new WebTarget with the query parameters added
  return newTarget;
}
```
We'd like to change it to use `queryParamsFromMap` like this:

```java
static WebTarget targetWithQueryParams(WebTarget webTarget, Map<String, Object> parameters) {
   // if we had an "escape hatch" we could do something like
   return WebTargetHelper.withWebTarget(webTarget).queryParamsFromMap(parameters).escape();
}
```

`WebTargetHelper` actually does have an "escape hatch" method named `wrapped()` which returns a `WebTarget`, but it is currently package-private. One easy option is to simply make it public. Does this make sense? Is `wrapped()` a good name for the escape hatch, or is there a better name?

**Comment by sleberknight** [comment]
The method could simply be named `toWebTarget()`. 

So then it would look like:

```Java
return WebTargetHelper.withWebTarget(target)
        .queryParamsFromMap(parameters)
        .toWebTarget();
```

**Comment by chrisrohr** [comment]
I like this idea!  Maybe make `wrapped()` public and rename it `toWebTarget()` as you mentioned.

**Comment by sleberknight** [comment]
I looked into this some more, and found that the existing package-private `wrapped()` method returns (as the method name indicates) the exact wrapped `WebTarget` instance, i.e. it simply does `return webTarget`. This is fine for the purpose it was used for, which is in the test, but I am not sure it's a good idea to just rename to `toWebTarget` and return the same exact `WebTarget` instance.

To me, `toWebTarget` implies that it returns a _new_ `WebTarget` instance that is equal to but not the same instance. Or put another way, do we really want to break the encapsulation by returning the same instance and possibly causing other problems due to callers having references to the wrapped `WebTarget`? And unfortunately, `WebTarget` is an interface and doesn't provide a way to create a copy of a `WebTarget`.

So, we could resolve this in a few ways that I can see. First, just leave the name `wrapped` and make that method public and update the Javadocs to indicate it returns the wrapped `WebTarget` instance and put the responsibility on the caller to treat it as a terminal operation and not use the `WebTargetHelper` after calling `wrapped()`. Second, rename to `toWebTarget()` and return a _new_ `WebTarget` instance, assuming we can find a not-completely-hacky way to do that. If we go with the second option, a bunch of tests will actually break, because they use the `wrapped()` method to make assertions, e.g. assert that we get the same instance back if the `queryParamIfNotBlank` method is passed a blank value. So, those tests would all need to be updated.

**Reply by sleberknight** [reply]
A third option I just thought of is to keep the existing package-private `wrapped()` method as-is, so the tests can continue working, but provide a new `toWebTarget()` method that returns a _new_ `WebTarget` instance (again, assuming we can find a not-too--hackish way to create that new instance).

**Reply by sleberknight** [reply]
This was fixed by #605
